Mathlib/RingTheory/DedekindDomain/FiniteAdeleRing.lean

```diff
@@ -30,450 +26,127 @@ field, its finite adèle ring is just defined to be the trivial ring.
-    Subalgebra.instFaithfulSMulSubtypeMem (subalgebra _ _)
-
-instance : FaithfulSMul (R_hat R K) (FiniteAdeleRing R K) :=
-  let h := FaithfulSMul.algebraMap_injective (R_hat R K) (K_hat R K)
-  (faithfulSMul_iff_algebraMap_injective _ _).2 ((funext (coe_algebraMap' R K)) ▸ h).of_comp
-
 section Topology
 
 open nonZeroDivisors
 open scoped Multiplicative
 
-variable {R K} in
-lemma mul_nonZeroDivisor_mem_finiteIntegralAdeles (a : FiniteAdeleRing R K) :
-    ∃ (b : R⁰) (c : R_hat R K), a * ((b : R) : FiniteAdeleRing R K) = c := by
-  let S := {v | a v ∉ adicCompletionIntegers K v}
-  choose b hb h using adicCompletion.mul_nonZeroDivisor_mem_adicCompletionIntegers (R := R) (K := K)
-  let p := ∏ᶠ v ∈ S, b v (a v)
-  have hp : p ∈ R⁰ := finprod_mem_induction (· ∈ R⁰) (one_mem _) (fun _ _ => mul_mem) <|
-    fun _ _ ↦ hb _ _
-  use ⟨p, hp⟩
-  rw [exists_finiteIntegralAdele_iff]
-  intro v
-  by_cases hv : a v ∈ adicCompletionIntegers K v
-  · exact mul_mem hv <| coe_mem_adicCompletionIntegers _ _
-  · dsimp only
-    have pprod : p = b v (a v) * ∏ᶠ w ∈ S \ {v}, b w (a w) := by
-      rw [← finprod_mem_singleton (a := v) (f := fun v ↦ b v (a v)),
-        finprod_mem_mul_diff (singleton_subset_iff.2 ‹v ∈ S›) a.2]
-    rw [pprod]
-    push_cast
-    rw [← mul_assoc]
-    exact mul_mem (h v (a v)) <| coe_mem_adicCompletionIntegers _ _
-
-theorem submodulesRingBasis : SubmodulesRingBasis
-    (fun (r : R⁰) ↦ Submodule.span (R_hat R K) {((r : R) : FiniteAdeleRing R K)}) where
-  inter i j := ⟨i * j, by
-    push_cast
-    simp only [le_inf_iff, Submodule.span_singleton_le_iff_mem, Submodule.mem_span_singleton]
-    exact ⟨⟨((j : R) : R_hat R K), by rw [mul_comm]; rfl⟩, ⟨((i : R) : R_hat R K), rfl⟩⟩⟩
-  leftMul a r := by
-    rcases mul_nonZeroDivisor_mem_finiteIntegralAdeles a with ⟨b, c, h⟩
-    use r * b
-    rintro x ⟨m, hm, rfl⟩
-    simp only [Submonoid.coe_mul, SetLike.mem_coe] at hm
-    rw [Submodule.mem_span_singleton] at hm ⊢
-    rcases hm with ⟨n, rfl⟩
-    simp only [LinearMapClass.map_smul, DistribMulAction.toLinearMap_apply, smul_eq_mul]
-    use n * c
-    push_cast
-    rw [mul_left_comm, h, mul_comm _ (c : FiniteAdeleRing R K),
-      Algebra.smul_def', Algebra.smul_def', ← mul_assoc]
-    rfl
-  mul r := ⟨r, by
-    intro x hx
-    rw [mem_mul] at hx
-    rcases hx with ⟨a, ha, b, hb, rfl⟩
-    simp only [SetLike.mem_coe, Submodule.mem_span_singleton] at ha hb ⊢
-    rcases ha with ⟨m, rfl⟩
-    rcases hb with ⟨n, rfl⟩
-    use m * n * (r : R)
-    simp only [Algebra.smul_def', map_mul]
-    rw [mul_mul_mul_comm, mul_assoc]
-    rfl⟩
-
-instance : TopologicalSpace (FiniteAdeleRing R K) :=
-  SubmodulesRingBasis.topology (submodulesRingBasis R K)
-
--- the point of `submodulesRingBasis` above: this now works
-example : IsTopologicalRing (FiniteAdeleRing R K) := inferInstance
+instance : IsTopologicalRing (FiniteAdeleRing R K) :=
+    haveI : Fact (∀ (v : HeightOneSpectrum R),
```

Comment:
I'm not a fan of `Fact`. Is there no way of avoiding it here? (Probably not strictly germane to this PR, more a question about how `RestrictedProduct.isTopologicalRing` is implemented)

Reply:
Yes, there's not a lot I can do about this here; `RestrictedProduct.isTopologicalRing` currently wants `[hBopen : Fact (∀ (i : ι), IsOpen ↑(B i))]` because typeclass inference won't trigger unless it can find the fact that the subgroups are open. I note that in `Topology/Algebra/RestrictedProduct.lean` we currently have
```
-- TODO: Make `IsOpen` a class like `IsClosed` ?
variable [hBopen : Fact (∀ i, IsOpen (B i : Set (R i)))]
```
so that's perhaps the way to proceed here, although I would imagine that the topologists would have opinions about this.